man/tunes2pod.1
.TH TUNES2POD "1" "June 2004" "tunes2pod.pl 0.95" "User Commands"
.SH NAME
tunes2pod.pl \- manual page for tunes2pod.pl 0.95
.SH SYNOPSIS
.B tunes2pod.pl
[\fIOPTION\fR]...
.SH DESCRIPTION
Convert an iTunesDB into XML (aka GNUtunesDB.xml)
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.TP
\fB\-m\fR, \fB\-\-mount=DIRECTORY\fR
iPod mountpoint. Default is $IPOD_MOUNTPOINT
.TP
\fB\-\-force\fR
Disable 'sync' checking. tunes2pod.pl may do stupid things :)
.SH AUTHOR
Written by Adrian Ulrich
.SH "REPORTING BUGS"
Report bugs to <bug-gnupod@nongnu.org>
.SH COPYRIGHT
Copyright \(co Adrian Ulrich
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B gnupod
is maintained as a Texinfo manual.  If the
.B info program
and
.B GNUpod
are properly installed at your site, the command
.IP
.B info gnupod
.PP
should give you access to the complete manual.


CHANGES
GNUpod 0.98.2 - codename "Jigen"

Summary
-------

- Changed the file-detection-type order
  of FileMagic.pm from:
  NonNative > QT > MP3 > PCM
  ..into..
  NonNative > QT > PCM > MP*

  ..MP3::Info had some false-positives and detected
  a few PCM-Files as MP3s :-/. The PCM-Detection
  is much simpler and shouldn't create any false-
  positives.
  Thanks to Stefano Vesa for reporting the Problem.

- GNUpod had some problems if the $IPOD_MOUNTPOINT
  contained whitespaces.
  Switched to File::Glob::bsd_glob() to fix this issue.

- Apple did it again! The iPodMini-FW1.4 doesn't like
  if length(path) > 55 .. rgs! 
  This version of GNUpod crops after 49 chars (to be on
  the safe side. Maybe Apple ...)
  Run 'gnupod_check.pl --fixit' to fix this issue (Maybe)

- gnupod_check.pl has now a --fixit switch. It will try to
  fix the found errors itself

- Updated documentation: Tell how to extract a
  Firmware-Image on Linux. (And others? Well, you need
  HFS+ Support)

  

*** OLD STUFF ***

GNUpod 0.98.1 - codename 'Lupin the 3rd'

Summary
-------

- Added support for multiple On-The-Go playlists.

- gnupod_addsong.pl can now reenocde aac and mp3 'on-the-fly'
  Usage: gnupod_addsong.pl --reencode 6 *.mp3
  This would reenocde everything with 'lame -V 6'
  Use this if you don't have much space left (iPod Shuffle?!)
  Note: The Reencode option also changes the quality of
        files converted with '--decode=' :-)

- Added support for m4a 'gnre' tags

- Fixed some iPod-Shuffle issues
   * mktunes.pl removes the iTunesShuffle leftover

- Remove TARGET-File if File::Copy() failed for any reason
   => Better handling of 'disk full' but still not perfect
      [GNUpod should write a 'sparefile' or something like that]


GNUpod 0.98 - codename 'Luna 9'

Summary
-------

- Support for iPod-Shuffle!
  (I do not know why Apple uses
   another Database for this thing!!!)
   Thanks to:
     * Chris Scott for testing and providing samples
     * Jeff Thieleke (From FooPod)
     * ..and Rodja Trappe (Samples + Testing)

- Fixed 'songnum' for ID3V1 Tags .. ouch

- Workaround for perl5.6 feature/bug
  'addtime' shouldn't be below zero anymore

- Write more tags for m4a files..

- Internal Change: Removed the ugly _itop and _x86itop..
  I hope i didn't break anything ;)

- Switched from Audio::FLAC to Audio::FLAC::Header



GNUpod 0.97 - codename guezli

Summary
-------

- 0.96 didn't write 'tracknum' for OGG Files, fixed now..
  Thanks to Ed Wandasiewicz for reporting this bug

- Read 'tracknum' on ID3V2 Files with 'TRK' Field

- Handle ID3Tags with Charset info

- You can now add MIDI files ;-) (with timidity)

- Added --set-songnum option for gnupod_addsong.pl

- Rewritten ConverterCore.. shouldn't break anything...

- gnupod_search.pl can now RENAME/Edit tags.
  use --rename="artist=New artist Name"

- You can now sort playlists:
  <playlist .. sort="artist"> or (reverse)
  <playlist .. sort="reverse title">

- Do not generate filenames longer than 64 chars..
  looks like the iPod Photo has a problem
  (Thanks to Mathieu Chouquet-Stringer for reporting)

- Added tools/fwpatch.pl. Allows you to convert an
  EU-iPod into an US-iPod (Make it louder)


GNUpod 0.96 - codename 'kartoffelsaft'

Summary
-------

- Changed --decode switch from gnupod_addsong.pl
  You can now select the output format of --decode
  -> pcm(old default) / MP3 (using lame) / AAC (using FAAC)  
  Example: # gnupod_addsong.pl --decode=mp3 *.flac
  (Idea and original patch from <daniel at kefa.co.uk>)

- GNUpod can now also convert OGG-Vorbis files
  (Use --decode=XXX, same as FLAC)

- Added --playlist=STRING switch to gnupod_addsong.pl
  This creates the playlist 'STRING' and adds all songs
  passed to gnupod_addsong.pl to this playlist.
  (Thanks to Bob Cotton)

- Many changes for the M4A parser, should work better now
  (Searches the SOUND Stream!)

- Write tracknum for AAC/ALAC files (doh!)

- Support for Bookmarks (Just rename .m4a files into .m4b)

- Convert strange Genre-Strings in MP3 Tags ( (17) -> Rock )

- Support for 'Playlist IDs' (Useless feature ATM ;) )


GNUpod 0.95 - codename 'sockentier'


Summary
-------

- GNUpod has now a configuration file! Read the docs or/and
  see doc/gnupodrc.example

- Added funky SmartPlaylist creator, see doc/mkspl.html
  (Needs JavaScript + DOM => Mozilla)
  ** SMARTPLAYLIST SYNTAX CHANGED **

- gnupod_addsong.pl can now add FLAC files.
  'flac' needs to be in your $PATH and it's disabled
  per default. (Enable it using gnupod_addsong.pl --decode
  or set 'gnupod_addsong.decode=1' in your config)

- gnupod_addsong.pl can read ALAC (Apples Loseless Audio Codec)
  files (iTunes creates normal .m4a files)

- Polished Playcounts parser (no hardcoded offsets anymore)

- added '--disable-v2' and '--disable-v1' to
  gnupod_addsong.pl (Disable parsing of ID3TAG version X)

- Fixed small bug with .m4a files and added more fields.
  gnupod_addsong.pl can now read M4A/AAC files from www.allofmp3.com

- added tools/m4a-info.pl: Use this tool to read Title/Artist tags
  from .m4a files (not installed by 'make install')

- You can modify the output of gnupod_search.pl using --view
  (See 'gnupod_search.pl --help')

- gnupod_addsong.pl checks for wrong extensions
  (like a mp3-file called 'test.wav'). The iPod wouldn't play
  such a file.

- Fixed some SmartPlaylist stuff and added support for last
  playtime

- Added gnupod_check.pl, it check's for 'lost' files

- I changed the 'build' system. Btw, you can now uninstall
  GNUpod using 'make uninstall' .. but why? :)

- Added manual pages

- GNUpod warns about case-sensitive mounted filesystems



GNUpod 0.94 - codename 'aspartam'


Summary (hmm.. it's a long summary..):
--------------------------------------
- Nobody did buy me a 40gb iPod with Firmware 2.x

- I bought one myself

- gnupod_addsong.pl understands .m4a files
  (= AAC Files written by iTunes)
  I don't know if it works with DRM files (.m4p?)

src/tunes2pod.pl
###__PERLBIN__###
#  Copyright (C) 2002-2004 Adrian Ulrich <pab at blinkenlights.ch>
#  Part of the gnupod-tools collection
#
#  URL: http://www.gnu.org/software/gnupod/
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# iTunes and iPod are trademarks of Apple
#
# This product is not supported/written/published by Apple!

use strict;
use GNUpod::iTunesDB;
use GNUpod::XMLhelper;
use GNUpod::FooBar;
use Getopt::Long;

use vars qw(%opts);
$| = 1;


print "tunes2pod.pl Version ###__VERSION__### (C) Adrian Ulrich\n";

$opts{mount} = $ENV{IPOD_MOUNTPOINT};

GetOptions(\%opts, "version", "force", "help|h", "mount|m=s");
GNUpod::FooBar::GetConfig(\%opts, {mount=>'s', force=>'b', anapodworkaround=>'b'}, "tunes2pod");


usage() if $opts{help};
version() if $opts{version};



#Normal operation
converter();

sub converter {
$opts{_no_sync} = 1;
my $con = GNUpod::FooBar::connect(\%opts);
usage("$con->{status}\n") if $con->{status};

#We disabled all autosyncing (_no_sync set to 1), so we do a test
#ourself
if(!$opts{force} && !(GNUpod::FooBar::_itb_needs_sync($con))) {
 print "I don't think that you have to run tunes2pod.pl\n";
 print "The GNUtunesDB looks up-to-date\n";
 print "\n";
 print "If you think i'm wrong, use '$0 --force'\n";
 exit(1);
}



open(ITUNES, $con->{itunesdb}) or usage("Could not open $con->{itunesdb}");



#Check where the FILES and PLAYLIST part starts..
#..and how many files are in this iTunesDB
my @itinfo = GNUpod::iTunesDB::get_starts(*ITUNES);

if(!defined(@itinfo)) {
  warn "File '$con->{itunesdb}' is not an iTunesDB, wrong magic in header!\n";
  exit(1);
}

#Start of Tracklist
my $tracklist_pos   = $itinfo[1]->{start};
my $tracklist_childs = $itinfo[1]->{childs};

#Start of Playlist
my $pl_pos   =  $itinfo[2]->{start};
my $pl_childs = $itinfo[2]->{childs};


my $podcast_pos    = $itinfo[3]->{start};
my $podcast_childs = $itinfo[3]->{childs};

print "> Has $tracklist_childs songs";

#Get all files
my $href= undef;
my $ff = 0;
my %hout = ();
for(my $i=0;$i<$tracklist_childs;$i++) {
	#get the mhit + all child mhods
	($tracklist_pos,$href) = GNUpod::iTunesDB::get_mhits($tracklist_pos, *ITUNES);
	#Seek failed.. this shouldn't happen..  
	if($tracklist_pos == -1) {
		print STDERR "\n*** FATAL: Expected to find $tracklist_childs files,\n";
		print STDERR "*** but i failed to get nr. $i\n";
		print STDERR "*** Your iTunesDB maybe corrupt or you found\n";
		print STDERR "*** a bug in GNUpod. Please send this\n";
		print STDERR "*** iTunesDB to pab\@blinkenlights.ch\n\n";
		exit(1);
	}
	GNUpod::XMLhelper::mkfile({file=>$href});  
	$ff++;
}
#<files> part built
print STDOUT "\r> Found $ff files, ok\n";


#Now get each playlist
print STDOUT "> Found ".($pl_childs-1)." playlists:\n";
for(my $i=0;$i<$pl_childs;$i++) {
	($pl_pos, $href) = GNUpod::iTunesDB::get_pl($pl_pos, {nomplskip=> $opts{anapodworkaround} }, *ITUNES); #Get an mhyp + all child mhods
	if($pl_pos == -1) {
		print STDERR "*** FATAL: Expected to find $pl_childs playlists,\n";
		print STDERR "*** but i failed to get nr. $i\n";
		print STDERR "*** Your iTunesDB maybe corrupt or you found\n";
		print STDERR "*** a bug in GNUpod. Please send this\n";
		print STDERR "*** iTunesDB to pab\@blinkenlights.ch\n\n";
		print STDERR "!!! If you are an 'Anapod' user, try to set\n";
		print STDERR "!!!   tunes2pod.anapodworkaround=1\n";
		print STDERR "!!! inside ~/.gnupodrc and re-run the command.\n";
		exit(1);
	}
	next if $href->{type};         #Don't list the MPL
	
	$href->{name} = "NONAME" unless($href->{name}); #Don't create an empty pl
	if(ref($href->{splpref}) eq "HASH" && ref($href->{spldata}) eq "ARRAY") { #SPL Data present
		print ">> Smart-Playlist '$href->{name}' found\n";
		render_spl($href->{name},$href->{splpref}, $href->{spldata}, $href->{matchrule},
		           $href->{content}, $href->{plid});
	}
	elsif($href->{podcast} == 1) {
		print ">> $href->{name}-Playlist with ".int(@{$href->{content}})." songs skipped\n";
	}
	else { #Normal playlist  
		print ">> Playlist '$href->{name}' with ".int(@{$href->{content}})." songs\n";
		GNUpod::XMLhelper::addpl($href->{name}, {plid=>$href->{plid}});
		foreach(@{$href->{content}}) {
		my $plfh = ();
		$plfh->{add}->{id} = $_->{sid};
		GNUpod::XMLhelper::mkfile($plfh,{plname=>$href->{name}});
		}
	}
}


# Read podcast childs if this iTunesDB provides one..
if($podcast_childs > 0) {
	print STDOUT "> Converting Podcasts-Playlists ...\n";
	my $pcnref = ();
	for(my $i=0;$i<$podcast_childs;$i++) {
		($podcast_pos, $href) = GNUpod::iTunesDB::get_pl($podcast_pos,{}, *ITUNES);
		
		if($href->{podcast} != 1) {
			next;
		}
		else {
			foreach my $pccont (@{$href->{content}}) {
				if(defined($pccont->{subnaming})) {
					print ">> $href->{name}-Playlist '$pccont->{subnaming}'\n";
					$pcnref->{$pccont->{plid}} = $pccont->{subnaming};
					GNUpod::XMLhelper::addpl($pccont->{subnaming}, {podcast=>1});
				}
				else {
					my $plfh = {add=>{id => $pccont->{sid}}};
					my $pcplname = $pcnref->{$pccont->{podcast_group_ref}};
					
					unless(defined($pcplname)) {
						warn "Unable to add $pccont->{sid} to a playlist: no reference to group $pccont->{podcast_group_ref} found\n";
					}
					else {
						GNUpod::XMLhelper::mkfile($plfh,{plname=>$pcplname});
					}
				}
			}
		}
	}
}




GNUpod::XMLhelper::writexml($con);
GNUpod::FooBar::setsync_itunesdb($con);
GNUpod::FooBar::setvalid_otgdata($con);

#The iTunes is now set to clean .. maybe we have to
#update the otg..
$opts{_no_sync} = 0;
GNUpod::FooBar::connect(\%opts);

print STDOUT "\n Done\n";
close(ITUNES);
exit(0);
}



#######################################################
# create a spl
sub render_spl {
 my($name, $pref, $data, $mr, $content, $plid) = @_;
 my $of = undef;
 $of->{liveupdate} = $pref->{live};
 $of->{moselected} = $pref->{mos};
 $of->{matchany}   = $mr;
 $of->{limitsort} = $pref->{isort};
 $of->{limitval}  = $pref->{value};
 $of->{limititem} = $pref->{iitem};
 $of->{checkrule} = $pref->{checkrule};
 $of->{plid}       = $plid;
#create this playlist
GNUpod::XMLhelper::addspl($name, $of);

  foreach my $xr (@{$data}) { #Add spldata
    GNUpod::XMLhelper::mkfile({spl=>$xr}, {splname=>$name});
  }
  foreach my $cont (@{$content}) { #Add (old?) content
    GNUpod::XMLhelper::mkfile({splcont=>{id=>$cont->{sid}}}, {splname=>$name});
  }

}





sub usage {
my($rtxt) = @_;
die << "EOF";
$rtxt
Usage: tunes2pod.pl [-h] [-m directory]

   -h, --help              display this help and exit
       --version           output version information and exit
   -m, --mount=directory   iPod mountpoint, default is \$IPOD_MOUNTPOINT
       --force             Disable 'sync' checking

Report bugs to <bug-gnupod\@nongnu.org>
EOF
}

sub version {
die << "EOF";
tunes2pod.pl (gnupod) ###__VERSION__###
Copyright (C) Adrian Ulrich 2002-2004

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

EOF
}




